### DISCUSSION TITLE
It would be good to add more examples implementations using kaoto.

### DISCUSSION THREAD

**Discussion by FlightControl-User** [KaotoIO/kaoto#1006] [Ideas]
The current examples of using kaoto are too easy.
It would be great to have people contributing, building sample implementations illustrating:
- message data transformation, using different techniques.
- message format transformation, including marshalling.
  - header variables handling
  - body variables handling
  - how to implement data mappings
  - use data properties
  - ...
- content-based routing.
- correlation.
- usage of synchronous and asynchronous calls.
- restul apis calling and consumption.
- use queueing.
- how to use beans and do bean implementations.
...

**Comment by lordrip** [comment]
Hi @FlightControl-User, thanks for raising this. How would you envision these sample implementations? As entries in [the workshop list](https://kaoto.io/workshop/)? Or similar to the [Camel examples repository](https://github.com/apache/camel-examples)?

**Comment by FlightControl-User** [comment]
I would actually envision both of them, but first to add them in the camel examples repository. And then work out the example on the web page and put a link on the web page to the github example repository.

I'm looking into kaoto and think it has great potential. However, when it comes to scripting or create concrete flows, I notice things aren't always so clear. I do have camel experience through coding using DSL, but when it comes to doing similar things with kaoto I struggle.
Also, it would be good to have a more detailed kaoto _manual_.

Note that the example repo is today _empty_.

Sven

**Comment by lordrip** [comment]
> I'm looking into kaoto and think it has great potential. However, when it comes to scripting or create concrete flows, I notice things aren't always so clear. I do have camel experience through coding using DSL, but when it comes to doing similar things with kaoto I struggle.

Thanks for this piece of feedback, can we expand on this topic? Could you point out where the struggle comes from, so we could work through them?

**Reply by lordrip** [reply]
We could use this thread to expand on things that are no so clear.

**Comment by FlightControl-User** [comment]
Okay. Thanks.Verzonden vanaf Outlook voor AndroidVan: Ricardo M. ***@***.***>Verstuurd: vrijdag, april 12, 2024 4:11:02 p.m.Aan: KaotoIO/kaoto ***@***.***>CC: Sven Van de Velde ***@***.***>; Mention ***@***.***>Onderwerp: Re: [KaotoIO/kaoto] It would be good to add more examples implementations using kaoto. (Discussion #1006)
We could use this thread to expand on things that are no so clear.

**Comment by lhein** [comment]
If you would like to contribute to the workshops on the webpage that would be very welcome. You can find the existing ones here: https://github.com/KaotoIO/kaoto.io/tree/main/content/workshop. If that is too much work you can also raise an issue in that repository with a specific workshop usecase and maybe already some working integration. We can then take care and create the workshop for it.

**Comment by FlightControl-User** [comment]
I see a lot of benefits demonstrating use cases that cover important integration patterns using simple open source adapters and format readers/writers. With simple file i/o and standard marshalling, one can easily demonstrate the key capabilities of Camel, and how to implement it using kaoto.

I would really need some high-level guidance on how to implement a data transformation, based on unmarshalled data using an out-of-the-box adapter like SWIFT MX to an other format (like SWIFT MT) and marshal it again and write the output in a file. I'm struggling how to write a processor in kaoto and how to use the unmarshalled data contained in java objects to implement a mapping! It is totally unclear to me at this moment how this is working.

Can somebody please clarify this mechanism and how to use the kaoto UI to realize this? (just in bullet points).
For example, i got this simple use case:

![image](https://github.com/KaotoIO/kaoto/assets/13690775/34293b4a-ea0a-43a3-96a8-335ce10c0a89)

Here i'm unmarshalling using the swiftmx adapter to a java object model;

![image](https://github.com/KaotoIO/kaoto/assets/13690775/d82a182d-4244-4d32-a97a-331c96a93b63)

So if i now add a processor:

![image](https://github.com/KaotoIO/kaoto/assets/13690775/3978486c-351d-4a96-87d0-e0f10d8a2f56)

where do I edit the code of the processor to map the exchange object onto the underlying java object models and to map it?
I'm totally lost in how to do that.

![image](https://github.com/KaotoIO/kaoto/assets/13690775/e1c9f4ab-92f2-4ec0-b333-f70977af4701)

**Comment by lhein** [comment]
The parsing should be straight forward:
```yaml
- route:
    id: route-2910
    from:
      id: from-2447
      uri: timer:myTimer
      parameters:
        delay: 10000
      steps:
        - setBody:
            id: setBody-2175
            expression:
              constant:
                expression: |-
                  {1:F01TESTUS20AXXX0000000000}{2:I362GHUJBBXXXXXXN}{4:
                  :15A:
                  :20:444444-000001011
                  :22A:NEWT
                  :94A:AGNT
                  :22C:AAAAUSX001TEST20
                  :23A:FLOATFIXED/GROSS
                  :21N:444444
                  :30V:20090106
                  :30P:20100105
                  :82A:AAAAUSX0
                  :87A:AAAAUSX0
                  :15D:
                  :33F:USD5000000,
                  :30X:20090406
                  :37G:2,5
                  :37R:2,
                  :37M:4,5
                  :30F:20090706
                  :15E:
                  :18A:1
                  :30F:20090706
                  :32M:USD56875,
                  :57A:AAAAUSX0
                  AAAAUSX0
                  -}
        - unmarshal:
            id: unmarshal-4292
            swiftMt:
              writeInJson: true
        - to:
            id: to-1860
            uri: log:ggg
            parameters: {}
```

The process you used above needs a bean reference. So you would actually need to create such a bean and depending on what you plan to do you probably have to write a Java class. That is atm a weak spot as we lack proper ootb support and assistance for that in Kaoto.

**Comment by FlightControl-User** [comment]
I had already succeeded myself to parse a PACS.008 SWIFT MX message by unmarshalling and then logging the output. That is the easy part, but then to develop a transformation it is a bit more tedious, correct? So how to start to write a transformer using a BEAN approach. Where do i start, I will have to edit the code using the VSCODE java editor I guess, correct? The "Ref" field in the process would need a "bean" procedure reference or something, correct?

**Comment by FlightControl-User** [comment]
Okay, I have more info. Looking into the tool called karavan, i found that when creating a process component step, when you give the reference (the name of the function that will process the incoming exchange object in java), you can click on a "+" button that generates the JAVA class for you in a file and adds this file to the camel repository!

Such a feature is missing in kaoto! I'll add a feature request for that! It is a small thing, but it really helps novice users like me to understand what is going on, without having to search hundreds of web pages and youtube videos ...

![image](https://github.com/KaotoIO/kaoto/assets/13690775/fa2e4279-43f9-4125-aea4-36dbeb062e32)

**Comment by lhein** [comment]
Thank you, Sven. We will look at that.

**Comment by FlightControl-User** [comment]
So I've considered your request to help build some examples and documentation. Gradually, I will build nice examples while learning camel in more depth and also kaoto (and karavan). So I'm happy to accept a contribution status to the project. Suggest to approach this in two steps, to allow contributors like myself to add content:

1. We add examples in the github repository foreseen for this, in a dedicated branch. We ask for a review from one of you experts to see if some aspects need to be changed or added. Then, we publish the new example in the main branch.
2. We then create website documentation for this example (write it all out), and again go through a review process before things get published.

A contributor guide on the kaoto github main page could also be written to guide potential future contributions to the process.

Sven

**Comment by lhein** [comment]
I think we could use the https://github.com/KaotoIO/kaoto-examples repository for building an example collection. What do you think?

**Comment by lhein** [comment]
We started to fill the examples repository.

**Comment by FlightControl-User** [comment]
That sounds like a good plan. I'm happy to contribute on this repository. Great!

**Comment by FlightControl-User** [comment]
Currently building this:

![image](https://github.com/KaotoIO/kaoto/assets/13690775/c89a937b-e06a-484f-a792-5e7c7861af99)

Jbang source:

```
- route:
    id: route-3942
    from:
      id: from-4901
      uri: timer
      parameters:
        period: "2000"
        timerName: template
      steps:
        - to:
            id: to-2357
            uri: https
            parameters:
              httpUri: random-data-api.com/api/v2/banks
        - process:
            id: process-3028
            ref: Convert
        - log:
            id: log-1580
            message: ${body}
```

The process has a processor camel java routine attached, with the following source code:

```java
import org.apache.camel.BindToRegistry;
import org.apache.camel.Configuration;
import org.apache.camel.Exchange;
import org.apache.camel.Processor;

@Configuration
@BindToRegistry("Convert")
public class Convert implements Processor {

  public void process(Exchange exchange) throws Exception {
      exchange.getIn().setBody("Hello World");
  }
}
```
But I get the following error when running it:

```
2024-04-21 09:08:23.920  INFO 10808 --- [           main] mel.cli.connector.LocalCliConnector : Management from Camel JBang enabled
org.apache.camel.FailedToCreateRouteException: Failed to create route route-3942 at: >>> process[ref:Convert] <<< in route: Route(route-3942)[From[timer:template?period=2000] -> [To[ht... because of No bean could be found in the registry for: Convert of type: org.apache.camel.Processor
        at org.apache.camel.reifier.RouteReifier.doCreateRoute(RouteReifier.java:228)
        at org.apache.camel.reifier.RouteReifier.createRoute(RouteReifier.java:79)
        at org.apache.camel.impl.DefaultModelReifierFactory.createRoute(DefaultModelReifierFactory.java:49)
        at org.apache.camel.impl.DefaultCamelContext.startRouteDefinitions(DefaultCamelContext.java:728)
        at org.apache.camel.impl.DefaultCamelContext.startRouteDefinitions(DefaultCamelContext.java:594)
        at org.apache.camel.impl.engine.AbstractCamelContext.doInit(AbstractCamelContext.java:2385)
        at org.apache.camel.support.service.BaseService.init(BaseService.java:78)
        at org.apache.camel.impl.engine.AbstractCamelContext.init(AbstractCamelContext.java:2021)
        at org.apache.camel.support.service.BaseService.start(BaseService.java:105)
        at org.apache.camel.impl.engine.AbstractCamelContext.start(AbstractCamelContext.java:2040)
        at org.apache.camel.impl.DefaultCamelContext.start(DefaultCamelContext.java:212)
        at org.apache.camel.main.KameletMain.doStart(KameletMain.java:324)
        at org.apache.camel.support.service.BaseService.start(BaseService.java:113)
        at org.apache.camel.dsl.jbang.core.commands.Run.runKameletMain(Run.java:1071)
        at org.apache.camel.dsl.jbang.core.commands.Run.run(Run.java:793)
        at org.apache.camel.dsl.jbang.core.commands.Run.doCall(Run.java:297)
        at org.apache.camel.dsl.jbang.core.commands.CamelCommand.call(CamelCommand.java:71)
        at org.apache.camel.dsl.jbang.core.commands.CamelCommand.call(CamelCommand.java:37)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
        at picocli.CommandLine.execute(CommandLine.java:2170)
        at org.apache.camel.dsl.jbang.core.commands.CamelJBangMain.run(CamelJBangMain.java:151)
        at org.apache.camel.dsl.jbang.core.commands.CamelJBangMain.run(CamelJBangMain.java:58)
        at main.CamelJBang.main(CamelJBang.java:36)
Caused by: org.apache.camel.NoSuchBeanException: No bean could be found in the registry for: Convert of type: org.apache.camel.Processor
        at org.apache.camel.reifier.AbstractReifier.mandatoryLookup(AbstractReifier.java:149)
        at org.apache.camel.reifier.ProcessReifier.createProcessor(ProcessReifier.java:41)
        at org.apache.camel.reifier.ProcessorReifier.makeProcessor(ProcessorReifier.java:879)
        at org.apache.camel.reifier.ProcessorReifier.addRoutes(ProcessorReifier.java:619)
        at org.apache.camel.reifier.RouteReifier.doCreateRoute(RouteReifier.java:224)
        ... 28 more

 *  The terminal process "C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe -Command jbang '-Dcamel.jbang.version=4.4.0' camel@apache/camel run test.camel.yaml --dev --logging-level=info  " terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it. 
 ```

Any idea what could be the reason?

Sven

**Reply by lhein** [reply]
Try to do 'camel run * --dev'.

**Comment by lhein** [comment]
I can run your route that way.....
![grafik](https://github.com/KaotoIO/kaoto/assets/435853/55ff48b5-8ec1-4fe2-8feb-41df6ffb95ee)

**Comment by lhein** [comment]
Raised an enhancement request at https://github.com/KaotoIO/vscode-kaoto/issues/556
